sound/soc/sof/ipc4-topology.c

```diff
@@ -0,0 +1,836 @@
+//
+//
+
+#include <linux/bits.h>
+#include <linux/device.h>
+#include <linux/errno.h>
+#include <linux/firmware.h>
+#include <linux/workqueue.h>
+#include <sound/pcm_params.h>
+#include <sound/sof/topology.h>
+#include <uapi/sound/sof/tokens.h>
+#include "ops.h"
+#include "sof-audio.h"
+#include "sof-priv.h"
+#include "ipc4.h"
+#include "ipc4-topology.h"
+#include "topology-common.h"
+
+static const struct sof_topology_token ipc4_sched_tokens[] = {
+	{SOF_TKN_SCHED_LP_MODE, SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
+		offsetof(struct sof_ipc4_pipeline, lp_mode), 0}
+};
+
+/**
+ * sof_ipc4_comp_alloc - allocate and initialize buffer for a new component
+ * @swidget: pointer to struct snd_sof_widget containing extended data
+ * @ipc_size: IPC payload size that will be updated depending on valid
+ *  extended data.
+ * @index: ID of the pipeline the component belongs to
+ *
+ * Return: The pointer to the new allocated component, NULL if failed.
+ */
+static struct sof_ipc_comp *sof_ipc4_comp_alloc(struct snd_sof_widget *swidget,
+						size_t *ipc_size, int index)
+{
+	u8 nil_uuid[SOF_UUID_SIZE] = {0};
+	struct sof_ipc_comp *comp;
+	size_t total_size = *ipc_size;
+
+	/* only non-zero UUID is valid */
+	if (memcmp(&swidget->comp_ext, nil_uuid, SOF_UUID_SIZE))
+		total_size += sizeof(swidget->comp_ext);
+
+	comp = kzalloc(total_size, GFP_KERNEL);
+	if (!comp)
+		return NULL;
+
+	comp->id = swidget->comp_id;
+	comp->pipeline_id = index;
+	comp->core = swidget->core;
+
+	/* handle the extended data if needed */
+	if (total_size > *ipc_size) {
+		/* append extended data to the end of the component */
+		memcpy((u8 *)comp + *ipc_size, &swidget->comp_ext, sizeof(swidget->comp_ext));
+		comp->ext_data_length = sizeof(swidget->comp_ext);
+	}
+
+	/* update ipc_size and return */
+	*ipc_size = total_size;
+	return comp;
+}
+
+static int sof_ipc4_widget_load_dai(struct snd_soc_component *scomp, int index,
+				    struct snd_sof_widget *swidget,
+				    struct snd_soc_tplg_dapm_widget *tw,
+				    struct sof_ipc4_dai *ipc4_dai)
+{
+	struct snd_soc_tplg_private *private = &tw->priv;
+	struct sof_ipc_comp_dai *comp_dai;
+	size_t size = sizeof(*comp_dai);
+	int ret;
+
+	comp_dai = (struct sof_ipc_comp_dai *)
+		   sof_ipc4_comp_alloc(swidget, &size, index);
+	if (!comp_dai)
+		return -ENOMEM;
+
+	ret = sof_parse_topology_tokens(scomp, comp_dai, SOF_TOKEN_DAI,
+					private->array, private->size);
+	if (ret != 0)
+		goto finish;
+
+	dev_dbg(scomp->dev, "dai %s, type %d\n", swidget->widget->name, comp_dai->type);
+
+	if (ipc4_dai) {
+		ipc4_dai->dai.scomp = scomp;
+
+		/*
+		 * copy only the sof_ipc_comp_dai to avoid collapsing
+		 * the snd_sof_dai, the extended data is kept in the
+		 * snd_sof_widget.
+		 */
+		memcpy(&ipc4_dai->dai.comp_dai, comp_dai, sizeof(*comp_dai));
+	}
+
+finish:
+	kfree(comp_dai);
+	return ret;
+}
+
+/*
+ * PCM Topology
+ */
+
+static int sof_ipc4_widget_load_pcm(struct snd_soc_component *scomp, int index,
```

Comment:
I feel there's a lot of duplicaed code we can avoid here. Can you please address the duplication and why it is unavaoidable @RanderWang @bardliao

Reply:
@ranj063  yes, we can reuse more code.

Reply:
@ranj063 @RanderWang After revisiting the code, I think only few code should be split. I will restructure the code. Thanks.